### PR TITLE
DNM: OSD: op lock completion

### PR DESCRIPTION
**Pull request by myoungwon**
This PR is based on #9286 which introduced op lock completion.
I create a new PR from previous discussion and @athanatos code.

Signed-off-by: Myoungwon Oh omwmw@sk.com

### CODE REVIEW

**Comment by athanatos** [comment]
@myoungwon Can you ping me about this next Monday?

**Comment by myoungwon** [comment]
@athanatos Yes, when can i ping you? (via irc)

**Comment by athanatos** [comment]
@myoungwon Nah, just comment on this PR like that, it'll show up in my email.  I just don't have time to go through it this week is all.

**Comment by myoungwon** [comment]
@athanatos OK.

**Comment by myoungwon** [comment]
@athanatos This is ping comment for you.

**Comment by athanatos** [comment]
@myoungwon Sorry, I know I asked you to ping me.  This week, I need to focus on EC overwrites and next week I'll be on vacation.  Ping me again in two weeks

**Comment by myoungwon** [comment]
@athanatos Ping comment again.

**Comment by athanatos** [comment]
@myoungwon I'm sorry, I know I've put you off twice now, but I'm not really going to have time to look at this until after the kraken feature freeze.

**Comment by myoungwon** [comment]
@athanatos Could you take a look this PR if you have finished your work?

**Comment by majianpeng** [comment]
@myoungwon  I wanted to test  w/ #9286 . w/ #9286 it can compile but it can't create a cluster. It always pended in creating status. There is no obvious  error message or assert in the log of osd.  Have you seen this problem? Thanks!

**Comment by myoungwon** [comment]
@majianpeng #9286 code use pg lock completion (original scheme) at initial time, then change op lock completion. (#11188 is latest source) anyway, this code is tested in the case of MOSDOp (general Read, Write cases). Therefore, please use existing jewel files for jobs such as creating pgs .., then change binary files to patched jewel for testing general Read, Write.

@liewegas @athanatos Can you take a look this PR?

**Comment by myoungwon** [comment]
@tchaikov @liewegas  I want to know whether this PR can be reviewed or not. This PR introduced "op lock completion" scheme as described in doc/dev/osd_internals/op_lock_completion.rst. "op lock completion" scheme had been improved in this discussion (#9286) with Samuel just (but maybe he have quit maintaining Ceph?). You can also refer original PR (#9286). If you think this is worthy, please let me know. I will do rebase.

**Comment by tchaikov** [comment]
@myoungwon  sorry for the latency. i will take a look at your PR tomorrow.

File: doc/dev/osd_internals/op_lock_completion.rst [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+D. when osd receives MSG_OSD_REPOPREPLY or MSG_OSD_EC_WRITE_REPLY
+  1. repop queue lock -> 2. find out OP -> 3. Repop(OP) lock (acquire) -> 
+  4. check completion (applied or commit) -> 5. If all commit or 
+  apply are finished, send ack to client -> 6. Repop(OP) lock (release) -> 
+  7. _queue_for_completion() -> 8. repop queue lock (release)
+
+E. _queue_for_completion (using ShardedOpWQ threads)
+  1. PG lock (acquire) -> 2. do_completion() (iterate repop_queue and 
+  repop_sub_queue for completion and remove RepGather and RepSub from 
+  repop_queue and repop_sub_queue) -> 3. PG lock (release)
```
Comment (liewegas) [inline doc/dev/osd_internals/op_lock_completion.rst]: Reading this it looks like the lock ordering is:

<pre>
- pg lock
  - repop queue lock
    - repop lock
</pre>
The _queue_for_completion description makes it sound like pg lock is taken while repop queue lock is held, it's just doing requeue_front().

Overall this strategy sounds fine, though!  I'm just sorry we took too long and the rebase is going to suck. :(

**Comment by liewegas** [comment]
I think that something like this is definitely in order.  It's unfortunate that we didn't get this through before the sharded wq code got refactored in OSD.cc; the rebase of Sam's patches will be annoying.

I would start by just cleaning up the initial patches that revamp the queueing; we'll want to test and validate those separately from the completion part anyway.

**Comment by myoungwon** [comment]
@liewegas If so, I will wait for your work (or I can cleaning up Sam's patches and create a new PR for the master branch if you want). Thanks

**Comment by liewegas** [comment]
If you can tackle refreshing Sam's patches that would be very helpful!

**Comment by myoungwon** [comment]
@liewegas @tchaikov  Thanks for your response. I will start working.

**Pull request closed without merging**